instrumentation/spring-messaging/src/main/java/brave/spring/messaging/TracingChannelInterceptor.java

```diff
@@ -0,0 +1,113 @@
+package brave.spring.messaging;
+
+import brave.Span;
+import brave.Tracing;
+import brave.propagation.ThreadLocalSpan;
+import brave.propagation.TraceContext;
+import brave.propagation.TraceContextOrSamplingFlags;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.support.ChannelInterceptor;
+import org.springframework.messaging.support.ChannelInterceptorAdapter;
+import org.springframework.messaging.support.ExecutorChannelInterceptor;
+import org.springframework.messaging.support.GenericMessage;
+import org.springframework.messaging.support.MessageHeaderAccessor;
+
+/**
+ * This implementation uses {@link ThreadLocalSpan} to propagate context between callbacks. This
+ * is an alternative to {@code ThreadStatePropagationChannelInterceptor} which is less sensitive
+ * to message manipulation by other interceptors.
+ */
+public final class TracingChannelInterceptor extends ChannelInterceptorAdapter implements
+    ExecutorChannelInterceptor {
+
+  public static ChannelInterceptor create(Tracing tracing) {
+    return new TracingChannelInterceptor(tracing);
+  }
+
+  final Tracing tracing;
+  final ThreadLocalSpan threadLocalProducerSpan;
+  final ThreadLocalSpan threadLocalConsumerSpan;
+  final TraceContext.Injector<MessageHeaderAccessor> injector;
+  final TraceContext.Extractor<MessageHeaderAccessor> extractor;
+
+  @Autowired TracingChannelInterceptor(Tracing tracing) {
+    this.tracing = tracing;
+    this.threadLocalProducerSpan = ThreadLocalSpan.create(tracing.tracer());
+    this.threadLocalConsumerSpan = ThreadLocalSpan.create(tracing.tracer());
+    this.injector = tracing.propagation().injector(MessageHeaderPropagation.INSTANCE);
+    this.extractor = tracing.propagation().extractor(MessageHeaderPropagation.INSTANCE);
+  }
+
+  @Override public Message<?> preSend(Message<?> message, MessageChannel channel) {
+    MessageHeaderAccessor headers = MessageHeaderAccessor.getMutableAccessor(message);
+    startProducerSpan(headers);
+    return new GenericMessage<>(message.getPayload(), headers.getMessageHeaders());
+  }
+
+  @Override
+  public void afterSendCompletion(Message<?> message, MessageChannel channel, boolean sent,
+      Exception ex) {
+    finishSpan(Span.Kind.PRODUCER, ex);
+  }
+
+  @Override public Message<?> postReceive(Message<?> message, MessageChannel channel) {
+    MessageHeaderAccessor headers = MessageHeaderAccessor.getMutableAccessor(message);
+    startConsumerSpan(headers);
+    return new GenericMessage<>(message.getPayload(), headers.getMessageHeaders());
```

Comment:
I think this really fixes the use-case for the `QueueChannel`  from Spring Integration.
Although I would also use `setLeaveMutable(true)` to allow downstream flow to reuse our *mutable* message.

I guess the concern from the `DirectChannel` perspective cannot be fixed here and we even might not need to worry about it. See https://github.com/spring-cloud/spring-cloud-sleuth/issues/821.

The problem there that a `MessageHandler` is called directly from the `send()` call stack. So, `MessageHandler.handle()` is in between `preSend()` and `afterSendCompletion()`, therefore it is still the part of the `PRODUCER` span. If that is really fine from the tracing perspective and there is no reason to fine-grain it for the `CONSUMER` span around this `MH.handle()` just because we still in the same thread and, yeah, in that case the producer span will be closed after that internal consumer span, so, we might just need to document that is it an expected behavior do not have separate span for the `MH` when we use `DirectChannel`. 

On the other hand I see the situation like: `DirectChannel -> RequestReplyMessageHandler -> DirectChannel` and so on. In this case we have several `preSend()` and `afterSendCompletion()` in the same thread. Therefore we start several producer spans, where the next one will be as a child of the previous. That will looks as a regular callstack. Having this assumption I would say that it might be good idea to really  fine-grain it for that child consumer span as well...

WDYT?

Reply:
right now, there are separate thread locals for producer and consumer, but if indeed there are several layers of nesting in-process, this would break. This is one of the tricky parts which relies on callbacks to be unwound the same way they were wound (ex a after for every pre). Stack semantics could work if this is the case

Reply:
Well, it doesn't because as I said: the `MH` in `DirectChannel` is called in between `preSend` and `afterSendComplete`:
```
message = interceptors.preSend(message, this, interceptorStack);
...
sent = this.doSend(message, timeout);
...
interceptors.postSend(message, this, sent);
interceptors.afterSendCompletion(message, this, sent, null, interceptorStack);
```

As you see there is on `beforeHandle` and `afterMessageHandled` because such a `MessageChannel` is not an `Executor`-aware. That's how we don't have there `consumer` span and that is what was an original request in that Spring Cloud Stream issue. 

My idea on the matter to check the channel type and perform consumer span withing `preSend/afterSendCompletion` call stack. However that looks for me like an overlap the producer span and consumer anyway...

Another thought is like a `TracingRequestHandlerAdvice` implementation to apply exactly to the `MessageHandler.handleRequestMessage()` call. This way with the scenario like: `dc -> mh -> dc -> mh` we may definitely have properly stacked producer and consumer spans even if everything is still in a single thread.

Thoughts ?

Reply:
I don't quite see the alignment problem as yet, if stacks are in use even if there's a handle between the send.

I think all it needs is a test case to make me understand. Any chance you can give a code snippet for a dumb `dc -> mh -> dc -> mh` that I can ensure passes? You can link me a similar test in spring-cloud-stream if easier

Reply:
Ok. I will try to come up with something tomorrow, but bear in mind there is reason to overhead it with Spring Cloud Stream, because the problem is exactly on the `MessageChannel` level. So, simple unit test should be enough.